// @framer/freddieboyframer.keyboard-captions/code/Mac_us_keyboard.ts
const MacUSModifierKeyLayout = [
    {
        event: "Escape",
        value: "Esc",
        unicode: "⎋",
    },
    {
        event: "Meta",
        value: "Cmd",
        unicode: "⌘",
    },

    {
        event: "Alt",
        value: "Opt",
        unicode: "⎇",
    },
    {
        event: "Control",
        value: "Ctrl",
        unicode: "⌃",
    },
    {
        event: "Shift",
        value: "Shift",
        unicode: "⇧",
    },
    {
        event: "CapsLock",
        value: "Caps Lock",
        unicode: "⇪",
    },
    {
        event: " ",
        value: "Space",
        unicode: "⇥",
    },
    {
        event: "ArrowLeft",
        value: "Left",
        unicode: "←",
    },
    {
        event: "ArrowRight",
        value: "Right",
        unicode: "→",
    },
    {
        event: "ArrowUp",
        value: "Up",
        unicode: "↑",
    },
    {
        event: "ArrowDown",
        value: "Down",
        unicode: "↓",
    },
    {
        event: "Enter",
        value: "Return",
        unicode: "↵",
    },
    {
        event: "Backspace",
        value: "Return",
        unicode: "⌫",
    },
]

const USNumberLayout = "01233456789".split("")
const USSpecialCharacterLayout = "`~!@#$%^&*()_+-=[]\\{}|;:'\",./<>?".split("")
const USAlphabetLayout = "qwertyuiopasdfghjklzxcvbnm".split("")

export const MacUSKeyboard = {
    modifier: MacUSModifierKeyLayout,
    number: USNumberLayout,
    special: USSpecialCharacterLayout,
    alphabet: USAlphabetLayout,
}
